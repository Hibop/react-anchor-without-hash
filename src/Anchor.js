import React from 'react'
import uniqid from 'uniqid'
import queryString from 'query-string'

const prefix = 'anchor-'

const getSearch = () => {
  const { location } = window

  const result = location.href.split('?')[1]

  if(result) {
    return `?${result}`
  }
}

class Anchor extends React.Component {
  constructor(props) {
    super(props)

    this.id = uniqid(prefix)
    this.handleHashChange = this.handleHashChange.bind(this)
    this.scroll = this.scroll.bind(this)
  }

  componentDidMount() {
    if ('scrollRestoration' in history) {
      history.scrollRestoration = 'manual'
    }
    window.addEventListener("hashchange", this.handleHashChange)
    this.scroll()
  }

  componentWillUnmount() {
    if ('scrollRestoration' in history) {
      history.scrollRestoration = 'auto'
    }
    window.removeEventListener("hashchange", this.handleHashChange)
  }

  scroll() {
    const { name, anchorKey, scrollIntoViewOption } = this.props
    const parsed = queryString.parse(getSearch())

    if (parsed && name === parsed[anchorKey]) {
      const dom = document.getElementById(this.id)

      if (dom.scrollIntoView) {
        dom.scrollIntoView(scrollIntoViewOption)
      }
    }
  }

  handleHashChange() {
    this.scroll()
  }

  render() {
    const { children } = this.props

    return (
      <div id={this.id}>
        {children}
      </div>
    )
  }
}

Anchor.defaultProps = {
  anchorKey: '_to',
  scrollIntoViewOption: true
}

export default Anchor