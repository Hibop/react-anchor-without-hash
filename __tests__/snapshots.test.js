import React from 'react'
import renderer from 'react-test-renderer'
import Anchor from '../src/Anchor'

describe('<Anchor/>', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<Anchor name="section1" />).toJSON()

    expect(tree).toMatchSnapshot()
  });

  test('renders with children correctly', () => {
    const tree = renderer.create(
      <Anchor name="section1">
        <div className="section section1">
          <h2>This is section1</h2>
          <div>There are some text...</div>
        </div>
      </Anchor>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  });

  test('renders with scrollTop correctly', () => {
    const anchorProps = {
      type: 'scrollTop',
      interval: 50
    }

    const tree = renderer.create(
      <Anchor name="section1" {...anchorProps}>
        <div className="section section1">
          <h2>This is section1</h2>
          <div>There are some text...</div>
        </div>
      </Anchor>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  });
});
