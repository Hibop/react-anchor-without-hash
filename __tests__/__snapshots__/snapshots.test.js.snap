// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anchor/> renders correctly 1`] = `<div />`;

exports[`<Anchor/> renders with children correctly 1`] = `
<div>
  <div
    className="section section1"
  >
    <h2>
      This is section1
    </h2>
    <div>
      There are some text...
    </div>
  </div>
</div>
`;

exports[`<Anchor/> renders with scrollTop correctly 1`] = `
<div>
  <div
    className="section section1"
  >
    <h2>
      This is section1
    </h2>
    <div>
      There are some text...
    </div>
  </div>
</div>
`;
